### PR TITLE
Handle down nodes instead of removing them in recovery

### PR DESCRIPTION
**Pull request by jrochas**

### CODE REVIEW

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java [hunk 1]
```diff
@@ -279,4 +267,81 @@ private boolean isEligible(RMNode node) {
...
+            return new Object[0];
+        }
+
+        @Override
+        public int getNumberOfActiveObjects() throws NodeException {
+            return 0;
+        }
+
+        @Override
+        public void killAllActiveObjects() throws NodeException, IOException {
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 2 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [NodeSource.java#L570](https://github.com/ow2-proactive/scheduling/blob/2980b3c5502c64e3381a1101edff00ffb4edbb55/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/NodeSource.java#L570): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RecoveryNode.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        return new Object[0];
+    }
+
+    @Override
+    public int getNumberOfActiveObjects() throws NodeException {
+        return 0;
+    }
+
+    @Override
+    public void killAllActiveObjects() throws NodeException, IOException {
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RecoveryNode.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java [hunk 3]
```diff
@@ -35,7 +37,13 @@
 import java.util.function.Function;
 
 import org.apache.log4j.Logger;
+import org.objectweb.proactive.ActiveObjectCreationException;
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java]: MINOR Remove this unused import 'org.objectweb.proactive.ActiveObjectCreationException'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java [hunk 4]
```diff
@@ -35,7 +37,13 @@
...
+import org.objectweb.proactive.core.ProActiveException;
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java]: MINOR Remove this unused import 'org.objectweb.proactive.core.ProActiveException'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java [hunk 5]
```diff
@@ -35,7 +37,13 @@
...
 import org.objectweb.proactive.core.node.Node;
+import org.objectweb.proactive.core.node.NodeException;
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java]: MINOR Remove this unused import 'org.objectweb.proactive.core.node.NodeException'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java [hunk 6]
```diff
@@ -25,6 +25,8 @@
  */
 package org.ow2.proactive.resourcemanager.core;
 
+import java.io.IOException;
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java]: MINOR Remove this unused import 'java.io.IOException'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java [hunk 7]
```diff
@@ -35,7 +37,13 @@
 import java.util.function.Function;
 
 import org.apache.log4j.Logger;
+import org.objectweb.proactive.ActiveObjectCreationException;
+import org.objectweb.proactive.core.ProActiveException;
 import org.objectweb.proactive.core.node.Node;
+import org.objectweb.proactive.core.node.NodeException;
+import org.objectweb.proactive.core.node.NodeInformation;
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java]: MINOR Remove this unused import 'org.objectweb.proactive.core.node.NodeInformation'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java [hunk 8]
```diff
@@ -25,6 +25,8 @@
  */
 package org.ow2.proactive.resourcemanager.core;
 
+import java.io.IOException;
+import java.io.Serializable;
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java]: MINOR Remove this unused import 'java.io.Serializable'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java [hunk 9]
```diff
@@ -35,7 +37,13 @@
 import java.util.function.Function;
 
 import org.apache.log4j.Logger;
+import org.objectweb.proactive.ActiveObjectCreationException;
+import org.objectweb.proactive.core.ProActiveException;
 import org.objectweb.proactive.core.node.Node;
+import org.objectweb.proactive.core.node.NodeException;
+import org.objectweb.proactive.core.node.NodeInformation;
+import org.objectweb.proactive.core.runtime.ProActiveRuntime;
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java]: MINOR Remove this unused import 'org.objectweb.proactive.core.runtime.ProActiveRuntime'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java [hunk 10]
```diff
@@ -35,7 +37,13 @@
...
+import org.objectweb.proactive.core.runtime.VMInformation;
```
Comment (sonarqube-activeeon) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java]: MINOR Remove this unused import 'org.objectweb.proactive.core.runtime.VMInformation'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [NodesLockRestorationManager.java#L83](https://github.com/ow2-proactive/scheduling/blob/28ff42b53b548c80787e23fd4da0a6b4c6199da1/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/recovery/NodesLockRestorationManager.java#L83): A "NullPointerException" could be thrown; "stopwatch" is nullable here. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)
1. ![MAJOR][MAJOR] [NodesRecoveryManager.java#L172](https://github.com/ow2-proactive/scheduling/blob/28ff42b53b548c80787e23fd4da0a6b4c6199da1/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/recovery/NodesRecoveryManager.java#L172): Catch Exception instead of Throwable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)
1. ![MAJOR][MAJOR] [NodeSource.java#L570](https://github.com/ow2-proactive/scheduling/blob/28ff42b53b548c80787e23fd4da0a6b4c6199da1/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/NodeSource.java#L570): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/recovery/RecoveryNode.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @Override
+    public Object setProperty(String s, String s1) throws ProActiveException {
+        return null;
+    }
+
+    @Override
+    public String getProperty(String s) throws ProActiveException {
+        return "";
```
Comment (fviale) [inline rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/recovery/RecoveryNode.java]: why not null ?

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [NodesLockRestorationManager.java#L83](https://github.com/ow2-proactive/scheduling/blob/fcf0777f1fa72066fdd717cdc0d2d9bd3722fc08/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/recovery/NodesLockRestorationManager.java#L83): A "NullPointerException" could be thrown; "stopwatch" is nullable here. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)
1. ![MAJOR][MAJOR] [NodesRecoveryManager.java#L172](https://github.com/ow2-proactive/scheduling/blob/fcf0777f1fa72066fdd717cdc0d2d9bd3722fc08/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/recovery/NodesRecoveryManager.java#L172): Catch Exception instead of Throwable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)
1. ![MAJOR][MAJOR] [NodeSource.java#L570](https://github.com/ow2-proactive/scheduling/blob/fcf0777f1fa72066fdd717cdc0d2d9bd3722fc08/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/NodeSource.java#L570): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [NodesLockRestorationManager.java#L83](https://github.com/ow2-proactive/scheduling/blob/8f5f51006e12811498206a6bb57f7ce3b53eab11/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/recovery/NodesLockRestorationManager.java#L83): A "NullPointerException" could be thrown; "stopwatch" is nullable here. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)
1. ![MAJOR][MAJOR] [NodesRecoveryManager.java#L172](https://github.com/ow2-proactive/scheduling/blob/8f5f51006e12811498206a6bb57f7ce3b53eab11/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/recovery/NodesRecoveryManager.java#L172): Catch Exception instead of Throwable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)
1. ![MAJOR][MAJOR] [NodeSource.java#L570](https://github.com/ow2-proactive/scheduling/blob/8f5f51006e12811498206a6bb57f7ce3b53eab11/rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/NodeSource.java#L570): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**